src/lib/tflm/tflm_backend.cc

```diff
@@ -192,38 +193,58 @@ static void ml_backend_init_ops_resolver(MicroOpsResolver *resolver) {
 int ml_backend_init_model(py_ml_model_obj_t *model) {
     RegisterDebugLogCallback(ml_backend_log_handler);
 
-    // Parse model's data.
+    // Parse the model's data.
     const Model *tflite_model = GetModel(model->data);
     if (tflite_model->version() != TFLITE_SCHEMA_VERSION) {
         mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Unsupported model schema"));
     }
 
-    // Initialize the op resolver.
+    // Initialize a temporary op resolver.
     MicroOpsResolver resolver;
     ml_backend_init_ops_resolver(&resolver);
 
-    // Allocate the interpreter and tensors once to initialize the model, check input
-    // and output data types and to get the optimal tensor arena size.
-    fb_alloc_mark();
-    uint32_t tensor_arena_size;
-    uint8_t *tensor_arena = (uint8_t *) fb_alloc_all(&tensor_arena_size, FB_ALLOC_PREFER_SIZE | FB_ALLOC_CACHE_ALIGN);
-
-    MicroInterpreter interpreter(tflite_model,
-                                 resolver,
-                                 tensor_arena,
-                                 tensor_arena_size);
+    gc_info_t info;
+    gc_info(&info);
+    // Allocate a temporary interpreter to get the optimal arena size.
+    size_t arena_size = info.max_free - TF_ARENA_ALIGN;
+    uint8_t *arena_memory = m_new(uint8_t, arena_size);
+    MicroInterpreter interpreter(tflite_model, resolver, arena_memory, arena_size);
     if (interpreter.AllocateTensors() != kTfLiteOk) {
         mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Failed to allocate tensors"));
     }
+    // Round up the optimal arena size to a multiple of the alignment.
+    arena_size = TF_ARENA_ROUND(interpreter.arena_used_bytes());
+    m_free(arena_memory);
+
+    // Allocate the persistent model state and interpreter.
+    ml_backend_state_t *state = m_new0(ml_backend_state_t, 1);
+    state->model = GetModel(model->data);
+    state->arena = m_new(char, arena_size + TF_ARENA_ALIGN);
+    state->resolver = new(m_new0(MicroOpsResolver, 1)) MicroOpsResolver();
+    ml_backend_init_ops_resolver(state->resolver);
+    uint8_t *aligned_arena = (uint8_t *) TF_ARENA_ROUND((uintptr_t) state->arena);
+    state->interpreter = new(m_new0(MicroInterpreter, 1)) MicroInterpreter(state->model,
+                                                                           *state->resolver,
+                                                                           aligned_arena,
+                                                                           arena_size);
+    if (state->interpreter->AllocateTensors() != kTfLiteOk) {
+        mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Failed to allocate tensors"));
+    }
+
+    // Initialize the model's state.
+    model->state = state;
+    model->memory_addr = (uint32_t) state->arena;
+    model->memory_size = arena_size;
```

Comment:
The extra 1KB was added originally because the returned value of RAM required for the tensorarea was sometimes less than what was actually needed. This may not be a problem anymore though.

Reply:
It was never was needed.